Tests/ApolloWebsocketTests/StarWarsSubscriptionTests.swift

```diff
@@ -252,4 +254,107 @@ class StarWarsSubscriptionTests: XCTestCase {
     subJedi.cancel()
     subNewHope.cancel()
   }
+  
+  // MARK: Data races tests
+  
+  func testConcurrentSubscribing() {
+    let firstSubscription = ReviewAddedSubscription(episode: .empire)
+    let secondSubscription = ReviewAddedSubscription(episode: .empire)
+    
+    let expectation = self.expectation(description: "Subscribers connected and received events")
+    expectation.expectedFulfillmentCount = 2
+    
+    var sub1: Cancellable?
+    var sub2: Cancellable?
+    
+    let queue = DispatchQueue(label: "com.apollographql.testing", attributes: .concurrent)
+    
+    queue.async {
+      sub1 = self.client.subscribe(subscription: firstSubscription) { _ in
+        expectation.fulfill()
+      }
+    }
+    
+    queue.async {
+      sub2 = self.client.subscribe(subscription: secondSubscription) { _ in
+        expectation.fulfill()
+      }
+    }
+    
+    // dispatched with a barrier flag to make sure
+    // this is performed after subscription calls
+    queue.sync(flags: .barrier) {
+      // dispatched on the processing queue to make sure
+      // this is performed after subscribers are processed
+      self.webSocketTransport.websocket.callbackQueue.async {
+        _ = self.client.perform(mutation: CreateReviewForEpisodeMutation(episode: .empire, review: ReviewInput(stars: 5, commentary: "The greatest movie ever!")))
+      }
+    }
+    
+    waitForExpectations(timeout: 10, handler: nil)
+    sub1?.cancel()
+    sub2?.cancel()
+  }
+  
+  func testConcurrentSubscriptionCancellations() {
+    let firstSubscription = ReviewAddedSubscription(episode: .empire)
+    let secondSubscription = ReviewAddedSubscription(episode: .empire)
+    
+    let expectation = self.expectation(description: "Subscriptions cancelled")
+    expectation.expectedFulfillmentCount = 2
+    
+    let sub1 = client.subscribe(subscription: firstSubscription) { _ in }
+    let sub2 = client.subscribe(subscription: secondSubscription) { _ in }
```

Comment:
Worth adding a post of a review, then failing either of these if something comes through since they're both being cancelled immediately?

Reply:
I can do that if you think it looks better, but I think it wouldn't really matter.

I feel like these tests only make sense if TSAN is turned on anyway.

Reply:
I'm not particularly concerned about the way it looks, but it'd be nice to have a way to test these without TSAN turn on since right now we're not able to turn it on for the library as a whole (yet)

Reply:
^still thinking on this - is this something where we should at least be able to say "These should never get called, even without TSAN on?'

Reply:
Yeah, I see what you mean. Kinda misunderstood you before.

I will perform a mutation after calling `cancel` to confirm.

Reply:
This is what I'm trying to do:  

    let firstSubscription = ReviewAddedSubscription(episode: .empire)
    let secondSubscription = ReviewAddedSubscription(episode: .empire)
    let expectation = self.expectation(description: "Subscriptions cancelled")
    expectation.expectedFulfillmentCount = 2
        
    let sub1 = client.subscribe(subscription: firstSubscription) { _ in
      XCTFail("Unexpected subscription response")
    }
    let sub2 = client.subscribe(subscription: secondSubscription) { _ in
      XCTFail("Unexpected subscription response")
    }
    
    let queue = DispatchQueue(label: "com.apollographql.testing", attributes: .concurrent)
    
    queue.async {
      sub1.cancel()
      expectation.fulfill()
    }
    queue.async {
      sub2.cancel()
      expectation.fulfill()
    }
    queue.async(flags: .barrier) {
      _ = self.client.perform(mutation: CreateReviewForEpisodeMutation(episode: .empire, review: ReviewInput(stars: 5, commentary: "The greatest movie ever!")))
    }
    
    waitForExpectations(timeout: 10, handler: nil)

But what happens is that final mutation sometimes triggers a subscription response in another test - `testMultipleSubscriptions`. Its callbacks get called too many times and the whole test fails. Testing this way can be achieved if the test sleeps after `waitForExpectations` for a while, but I feel thats just messy.

Reply:
I think waiting for expectations is totally reasonable - you also can do an inverted expectation where the test fails if it actually gets called - so instead of having the explicit `XCTFail`s you can just have an `invertedExpectation.fulfill()`. You'd probably only need to wait that one 1-2 seconds, since the perform mutation should happen a lot faster than that.